### PR TITLE
fix: multi-source app refresh (#11772)

### LINKED ISSUE #11772: 2.6 RC1: multi source application needs a hard refresh to deploy changes
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When creating a multi source application, argocd needs a hard refresh to deploy changes, when changes are applied in the sources.

**To Reproduce**
Create an application with helm source and values from git repo.

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: argo-rollouts
  namespace: argocd
  labels:
    type: deployment
    application: argo-rollouts
spec:
  project: default
  sources:
    - repoURL: 'https://argoproj.github.io/argo-helm'
      targetRevision: 2.21.1
      chart: argo-rollouts
      helm:
        valueFiles:
        - $values/argocd/applications/values/argo-rollouts/values-common.yaml
    - repoURL: https://github.com/xxx/yyy.git
      targetRevision: HEAD
      ref: values
  destination:
    server: 'https://kubernetes.default.svc'
    namespace: argo-rollouts
  syncPolicy:
    syncOptions:
      - CreateNamespace=true
    automated: {}
```
Once the app is deployed, upgrading the values in the git repo, does not trigger any change in argocd. 
In order to make argocd aware of this change, a hard-refresh is needed.

**Expected behavior**

Argocd controller should note that the repo is out of sync, without the need of a hard-refresh.

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Fixes #11772

This turns out to be gnarlier than I'd hoped.

Basically, the manifest cache for a source is currently based on _only_ the commit hash of the source, not any other sources it references.

I'm adding code to inject the other sources' commits in the cache key. 

This won't help webhooks. There's currently no way for a webhook from referenced-repo-Y to trigger a refresh on referencing-source-X. I'm leaving that fix for another PR.

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -434,11 +440,55 @@ func (s *Service) runRepoOperation(
...
+						refKeys := make([]string, 0)
+						for refKey := range refSources {
+							refKeys = append(refKeys, refKey)
+						}
+						return nil, fmt.Errorf("source referenced %q, which is not one of the available sources (%s)", refVar, strings.Join(refKeys, ", "))
+					}
+					if refSourceMapping.Chart != "" {
+						return nil, fmt.Errorf("source has a 'chart' field defined, but Helm charts are not yet not supported for 'ref' sources")
+					}
+					normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
```
Comment (ishitasequeira) [inline reposerver/repository/repository.go:470]: Should we move this to a new function as we are using the exact same steps [here](https://github.com/argoproj/argo-cd/blob/master/reposerver/repository/repository.go#L623-L642)? WDYT?

**Review by ishitasequeira — COMMENTED** [review]
Tested the PR locally and it is working as expected. Just a minor nit.

File: reposerver/repository/repository.go (hunk 1 above)
Comment (crenshaw-dev) [reply reposerver/repository/repository.go:472]: Per Slack convo, added comments to note the duplication. We can try to clean up in a follow-up refactoring PR after 2.6 is out.

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Review by leoluz — APPROVED** [review]
As discussed with @crenshaw-dev, there is room for improvement in the current PR to avoid adding an additional parameter to an already big method signature in `Cache.GetManifests`. The agreement was to add `TODO` comments in the code so we can unblock 2.6 release and address this refactoring later.
LGTM

File: reposerver/cache/cache.go [hunk 2]
```diff
@@ -212,8 +221,8 @@ func LogDebugManifestCacheKeyFields(message string, reason string, revision stri
 	}
 }
 
-func (c *Cache) GetManifests(revision string, appSrc *appv1.ApplicationSource, srcRefs appv1.RefTargetRevisionMapping, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string, res *CachedManifestResponse) error {
-	err := c.cache.GetItem(manifestCacheKey(revision, appSrc, srcRefs, namespace, trackingMethod, appLabelKey, appName, clusterInfo), res)
+func (c *Cache) GetManifests(revision string, appSrc *appv1.ApplicationSource, srcRefs appv1.RefTargetRevisionMapping, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string, res *CachedManifestResponse, refSourceCommitSHAs ResolvedRevisions) error {
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply reposerver/cache/cache.go:228]: Added a TODO to the code. We should refactor next time we have to fix a bug related to this code.

**Comment by crenshaw-dev** [comment]
Cherry-picked onto release-2.6 for 2.6.0.

**Comment by ls-julien-deruere** [comment]
Hi Folks, since this exact commit happened, we cannot upgrade our ArgoCD to more than 2.5.22. Our apps are synced but never (hard)refreshed first. It's happening only on one of our cluster, do you have an explanation?

**Pull request merged**